gbuiltin/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Helper crate defining Gear built-in actors communication protocol.
+//!
+//! This crate defines a set of types that contracts can use to interact
+//! with the so-called "builtin" actors - that is the actors that are defined
+//! for any Gear runtime and provide an API for the applications to build on top
+//! of some blockchain logic like staking, governance, etc.
+
+//! For a builtin actor to process a message, it should be able to decode its
+//! payload into one of the supported message types.
+//!
+//! # Examples
+//!
+//! The following example shows how a contract can send a message to a builtin actor
+//! (specifically, a staking actor) to bond some `value` to self as the controller
+//! so that the contract can later use the staking API to nominate validators.
+//!
+//! ```ignore
+//! use gstd::{msg, ActorId};
+//! use gbuiltin::staking::{Request, RequestV1, RewardAccount};
+//! use parity_scale_codec::Encode;
+//!
+//! const BUILT_IN: ActorId = ActorId::new(hex_literal::hex!(
+//!     "9d765baea1938d17096421e4f881af7dc4ce5c15bb5022f409fc0d6265d97c3a"
+//! ));
+//!
+//! #[gstd::async_main]
+//! async fn main() {
+//!     let value = msg::value();
+//!     let payee: Option<RewardAccount> = None;
+//!     let payload = Request::V1(RequestV1::Bond { value, payee }).encode();
+//!     let _ = msg::send_bytes_for_reply(BUILT_IN, &payload[..], 0, 0)
+//!         .expect("Error sending message")
+//!         .await;
+//! }
+//! # fn main() {}
+//! ```
+//!
+
+#![no_std]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+use scale_info::scale::{self, Decode, Encode};
+
+pub type AccountId = [u8; 32];
+
+pub use error::DispatchErrorReason;
+
+/// Message processing output
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+#[codec(crate = scale)]
+pub enum Response {
+    Success,
+    Failure(DispatchErrorReason),
+}
+
+impl<T> From<Result<T, DispatchErrorReason>> for Response {
+    fn from(result: Result<T, DispatchErrorReason>) -> Self {
+        result.map_or_else(Response::Failure, |_| Response::Success)
+    }
+}
+
+pub mod error {
+    use super::*;
+
+    /// Type representing a dispatched Runtime call internal error.
+    // TODO: see if we can add more granularity to the error type (e.g. describe the pallet etc.)
+    #[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, derive_more::Display)]
+    #[codec(crate = scale)]
+    pub enum DispatchErrorReason {
+        #[display(fmt = "Runtime internal error")]
+        RuntimeError,
+    }
+}
+
+// TODO: review and adjust
+pub mod staking {
```

Comment:
these should be deleted for now

Reply:
and qs to consider: in case of incompatibility of the api shouldn't we truly create a new builtin?

Reply:
This is indeed a better approach I think